mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs

```diff
@@ -216,5 +216,45 @@ void GetAvailableThreads_Callback (object state)
 
 			mre.WaitOne ();
 		}
+
+		[Test]
+		public void AsyncLocalCapture ()
+		{
+			var asyncLocal = new AsyncLocal<int>();
+			asyncLocal.Value = 1;
+			int var_0, var_1, var_2, var_3;
+			var_0 = var_1 = var_2 = var_3 = 99;
+			int total = 0;
+
+			var evt = new AutoResetEvent (false);
+			ThreadPool.QueueUserWorkItem(state => {
+				var_0 = asyncLocal.Value;
+				Interlocked.Increment (ref total);
+			}, null);
+
+			ThreadPool.UnsafeQueueUserWorkItem(state => {
+				var_1 = asyncLocal.Value;
+				Interlocked.Increment (ref total);
+			}, null);
+
+			ThreadPool.RegisterWaitForSingleObject (evt, (state, to) => {
+				var_2 = asyncLocal.Value;
+				Interlocked.Increment (ref total);
+			}, null, 1, false);
+
+			ThreadPool.UnsafeRegisterWaitForSingleObject (evt, (state, to) => {
+				var_3 = asyncLocal.Value;
+				Interlocked.Increment (ref total);
+			}, null, 1, false);
+
+			while (total < 4)
```

Comment:
I'd use a CountdownEvent with timeout as this block indefinitely on an error

Reply:
I'd use `ManuelResetEvent` for each case, so you can figure out which one has finished and which one hasn't. You can wait on all of them with a `Assert.IsTrue(WaitHandle.WaitAll(mres, 5000))`.